### ISSUE TITLE
Add support for FPM

### ISSUE THREAD

**Issue by gnikit** [fortran-lang/vscode-fortran-support#471]
The Fortran Package Manager is a package manager and build system for Fortran developed by us. Modern Fortran should natively support it, allow for easy project setup, etc.

**Comment by fedebenelli** [comment]
I'm really interested in this feature, while setting up tasks to interact with fpm ain't that hard, a full integration with vscode will surely be a good thing to ease the process. Some kind of integration with debugging would be nice, too. Right now I only know of three ways of debugging:

- Having a separate Makefile (with it's specific task) to compile the executable to a specific directory and keep a fixed `launch.json` file. 
- Building with `fpm` and moving the executable to a specific dir (to keep the `launch.json` file intact).
- Build with `fpm` and edit the `launch.json` file to point to the new executable.

Maybe there is a better way that I'm not aware of...

I think that good enough interaction with `fpm` to both run, debug and test would be awesome.

**Comment by gnikit** [comment]
Hi @fedebenelli thanks for the interest in this issue, I haven't had a lot of time to think about how we will best implement this. My original thoughts were to offer some high level integration for building targets (similar to CMake), you wouldn't need to explicitly set build tasks in `launch.json`.

For debugging I would suspect an out of source debug installation would be what you are looking for, so your debug targets would not affect your optimised targets. My workflow in general is me using debug mode and if I need to run with optimisations enabled I will reconfigure and recompile.

If you have any ideas or suggestions feel free to post them on this issue. Later this summer I will be opening a thread on [Fortran Discourse](https://fortran-lang.discourse.group/) to get some ideas about how the UI should look like.

**Comment by ivan-pi** [comment]
You may want to keep an eye on the discussion at https://github.com/fortran-lang/fpm/discussions/479

I've shown a basic example of fpm in Sublime (https://github.com/fortran-lang/fpm/discussions/479#discussioncomment-1778860) and @plevold has demonstrated it for VS Code (https://github.com/fortran-lang/fpm/discussions/479#discussioncomment-1238142). It works best if you have your code open as a "Project". In this case, you can easily run fpm in the top folder containing the manifest.

I think it's good to adopt a co-design strategy, meaning if there is something VS Code needs, we can see to add it in fpm.

**Comment by radekg** [comment]
Awesome to see this here. I have stumbled on the issue related to fpm + this extension: https://github.com/fortran-lang/fpm/issues/732.

It would be great if the Modern Fortran extension picked up the `build/` directory as the `mod` files directory when `fpm.toml` file is found in the project's root.

A question, in the meantime, is there a way to configure this `vsc` extension to always store mod files in `build/`, and pick them up from the same directory?

**Comment by gnikit** [comment]
You can set your global vscode setting call modOutputDir (I think) to be something like `${workspaceFolder}/build` that should both store your .mod files in build/ and include the path to the linter (gfortran should do that by default)

**Comment by radekg** [comment]
> You can set your global vscode setting call modOutputDir (I think) to be something like `${workspaceFolder}/build` that should both store your .mod files in build/ and include the path to the linter (gfortran should do that by default)

Mmmm, this is not going to be so easy. Apparently, I need to hard-code that `gfortran_...` directory, too.

![Screen Shot 2022-08-26 at 00 13 55](https://user-images.githubusercontent.com/62946/186778277-22015739-8ded-44fa-91e5-12535016fcab.png)

**Comment by gnikit** [comment]
Fpm can run additional commands to the shell by passing them after the fpm specific commands e.g. fpm build -- find build -name '*.mod' -exec cp {} ./build/include \;

You could define in a build task or a series of them that will accomplish what you want. I think I posted somewhere in the fpm discussions a series of tasks.json and launch.json files that get vscode to build and debug fpm built codes.

What you are effectively trying to do is what this issue is about. Working out that hash is part of the problem. There was a discussion in fpm about making the paths somewhat deterministic so tools would be able to access builds. 

Basically this issue is still on brainstorming stage about how to implement the vscode integration.

**Comment by gnikit** [comment]
Again, the additional command I posted will not work, since you might be copying .mod files from outdated or unrelated builds. The easiest way we found was to install the package and it's mod files into some temporary location for vscode to access.

**Comment by radekg** [comment]
Thank you, I will look into those when time allows. I’m learning Fortran for fun, pretty cool so far.

**Comment by gnikit** [comment]
I am glad to hear that, if you want any help don't hesitate to reach out here or in our Fortran Discourse https://fortran-lang.discourse.group/
